### PR TITLE
feat: deprecate input output

### PR DESCRIPTION
**Pull request by guilhem-barthes**
## Related issue

- https://github.com/Substra/substra-backend/pull/509
- https://github.com/Substra/substra-tests/pull/256
- https://github.com/Substra/substrafl/pull/129

## Summary

Modify substra to not rely on `asset` / `value` in Input assets and output assets.

Fixes FL-652

## Test

Ci in `substra` companion PR

## Please check if the PR fulfills these requirements

- [x] If necessary, the [changelog](https://github.com/Substra/substra/blob/main/CHANGELOG.md) has been updated
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [ ] The commit message follows the [conventional commit](https://www.conventionalcommits.org/en/v1.0.0/) specification
- For any breaking changes, companion PRs have been opened on the following repositories:
  - [ ] [substra-tests](https://github.com/Substra/substra)
  - [ ] [substrafl](https://github.com/Substra/substrafl)
  - [ ] [substra-documentation](https://github.com/Substra/substra-documentation)

### CODE REVIEW

**Comment by guilhem-barthes** [comment]
/e2e --refs substra-backend=feat/remove_deprecated_input_output_values

**Comment by Owlfred** [comment]
End to end tests: :hourglass_flowing_sand: **STARTING**

**Comment by guilhem-barthes** [comment]
/e2e --refs substra-backend=feat/remove_deprecated_input_output_values

**Comment by guilhem-barthes** [comment]
/e2e --refs substra-backend=feat/remove_deprecated_input_output_values,substra-tests=chore/remove-output.value-from-tests

**Comment by Owlfred** [comment]
End to end tests: :hourglass_flowing_sand: **STARTING**

**Comment by guilhem-barthes** [comment]
/e2e --refs substra-backend=feat/remove_deprecated_input_output_values,substra-tests=chore/remove-output.value-from-tests

**Comment by Owlfred** [comment]
End to end tests: :hourglass_flowing_sand: **STARTING**

**Comment by guilhem-barthes** [comment]
/e2e --refs substra-backend=feat/remove_deprecated_input_output_values,substra-tests=chore/remove-output.value-from-tests,substrafl=feat/remove-deprecated-input-output-values

**Comment by Owlfred** [comment]
End to end tests: :hourglass_flowing_sand: **STARTING**

**Comment by guilhem-barthes** [comment]
/e2e --refs substra-backend=feat/remove_deprecated_input_output_values,substra-tests=chore/remove-output.value-from-tests

**Comment by Owlfred** [comment]
End to end tests: :hourglass_flowing_sand: **STARTING**

**Comment by guilhem-barthes** [comment]
/e2e --refs substra-backend=feat/remove_deprecated_input_output_values,substra-tests=chore/remove-output.value-from-tests

**Comment by Owlfred** [comment]
End to end tests: :hourglass_flowing_sand: **STARTING**

**Comment by guilhem-barthes** [comment]
/e2e --refs substra-backend=feat/remove_deprecated_input_output_values,substra-tests=chore/remove-output.value-from-tests

**Comment by Owlfred** [comment]
End to end tests: :hourglass_flowing_sand: **STARTING**

**Comment by guilhem-barthes** [comment]
/e2e --tests=sdk,substrafl --benchmarks=camelyon --refs substra-backend=feat/remove_deprecated_input_output_values,substra-tests=chore/remove-output.value-from-tests

**Comment by Owlfred** [comment]
End to end tests: :hourglass_flowing_sand: **STARTING**

**Comment by guilhem-barthes** [comment]
/e2e --tests=sdk,substrafl --benchmarks=camelyon --refs substra-backend=feat/remove_deprecated_input_output_values,substra-tests=chore/remove-output.value-from-tests

**Comment by Owlfred** [comment]
End to end tests: :hourglass_flowing_sand: **STARTING**

**Comment by guilhem-barthes** [comment]
/e2e --tests=sdk,substrafl --benchmarks=camelyon --refs substra-backend=feat/remove_deprecated_input_output_values,substra-tests=chore/remove-output.value-from-tests

**Comment by Owlfred** [comment]
End to end tests: :hourglass_flowing_sand: **STARTING**

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -10,6 +10,7 @@ repos:
     hooks:
       - id: flake8
         additional_dependencies: [pep8-naming, flake8-bugbear]
+        args: ['--classmethod-decorators=classmethod,validator,root_validator']
```
Comment (SdgJlbl) [inline .pre-commit-config.yaml:13]: out of curiosity, what does this do?

File: CHANGELOG.md [hunk 2]
```diff
@@ -7,6 +7,19 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+### Added
+
+- `InputAsset` and `OutputAsset` ([#361](https://github.com/Substra/substra/pull/361))
```
Comment (SdgJlbl) [inline CHANGELOG.md:12]: I think it would be worth elaborating a bit ^^

File: substra/sdk/client.py [hunk 3]
```diff
@@ -797,7 +797,25 @@ def list_task(
 
 
         """
-        return self._list(schemas.Type.SummaryTask, filters, order_by, ascending)
+        return self._list(schemas.Type.Task, filters, order_by, ascending)
+
+    @logit
+    def list_task_input_assets(self, key: str) -> List[models.InputAsset]:
+        """List input assets for a specific task, the returned object is described
+        in the [models.Task](sdk_models.md#Task) model"""
```
Comment (ThibaultFy) [inline substra/sdk/client.py:805]:
```suggestion
        in the [models.InputAsset](sdk_models.md#InputAsset) model"""
```

File: substra/sdk/client.py [hunk 4]
```diff
@@ -797,7 +797,25 @@ def list_task(
...
+        return self._backend.list_task_input_assets(key)
+
+    @logit
+    def list_task_output_assets(self, key: str) -> List[models.OutputAsset]:
+        """List output assets for a specific task, the returned object is described
+        in the [models.Task](sdk_models.md#Task) model"""
```
Comment (ThibaultFy) [inline substra/sdk/client.py:811]:
```suggestion
        in the [models.OutputAsset](sdk_models.md#OutputAsset) model"""
```

File: substra/sdk/schemas.py [hunk 5]
```diff
@@ -54,13 +54,21 @@ class Type(enum.Enum):
...
         """Returns the name used to identify the asset on the backend."""
         name = self.value
         return _SERVER_NAMES.get(name, name)
 
+    def to_asset_kind(self):
+        """Returns the name used to identify the asset on the backend."""
+        if self.value == self.Dataset.value:
+            return AssetKind.data_manager
+        else:
+            return AssetKind[self.value]
```
Comment (ThibaultFy) [inline substra/sdk/schemas.py:70]:
```suggestion
```
No use

File: substra/sdk/backends/local/compute/worker.py [hunk 6]
```diff
@@ -196,6 +205,7 @@ def _prepare_datasamples_inputs_and_paths(
 
     def _save_output(
         self,
+        *,
```
Comment (SdgJlbl) [inline substra/sdk/backends/local/compute/worker.py:208]: 👍

**Review by ThibaultFy — COMMENTED** [review]
Thanks a lot !!!

File: .pre-commit-config.yaml (hunk 1 above)
Comment (guilhem-barthes) [reply .pre-commit-config.yaml:13]: Define what to be considered [as class method in `flake8`](https://github.com/PyCQA/pep8-naming/issues/38), making not mandatory to add `#noqa: N805` in front of each validator in pydantic (some are removed in this PR). By default, flake8 considers that the first argument has to be `self`, except if decorated with `classmethod`. This argument add `validator` and `root_validator` to this list.

File: substra/sdk/backends/local/compute/worker.py [hunk 7]
```diff
@@ -133,15 +135,22 @@ def _get_cmd_template_inputs_outputs(
...
-        in_task = self._db.get(schemas.Type.Task, task_input.parent_task_key)
-        input_artifact = in_task.outputs[task_input.parent_task_output_identifier].value
-        assert isinstance(
-            input_artifact, models.OutModel
-        ), "The task_input value must be an artifact, not a performance"
+    def _prepare_artifact_input(self, task_input, task_key, input_volume, multiple):
+        outputs = self._db.list(
+            schemas.Type.OutputAsset,
+            {"compute_task_key": task_input.parent_task_key, "identifier": task_input.parent_task_output_identifier},
+        )
```
[the remark this answers is not in the document]
Comment (guilhem-barthes) [reply substra/sdk/backends/local/compute/worker.py:142]: The client is not imported in the worker, in fact the client includes a worker

File: substra/sdk/schemas.py (hunk 5 above)
Comment (guilhem-barthes) [reply substra/sdk/schemas.py:70]: It's actually used there: https://github.com/Substra/substra/blob/41b9a24a76fec12fa84dc2bff62122ffae352481/substra/sdk/backends/local/compute/worker.py#L322

File: CHANGELOG.md (hunk 2 above)
Comment (guilhem-barthes) [reply CHANGELOG.md:12]: Which can of information about these do you think are relevant to mention?

**Comment by guilhem-barthes** [comment]
/e2e --tests=sdk,substrafl --benchmarks=camelyon --refs substra-backend=feat/remove_deprecated_input_output_values,substra-tests=chore/remove-output.value-from-tests

**Comment by Owlfred** [comment]
End to end tests: :hourglass_flowing_sand: **STARTING**

**Comment by guilhem-barthes** [comment]
/e2e --tests=sdk,substrafl --benchmarks=camelyon --refs substra-backend=feat/remove_deprecated_input_output_values,substra-tests=chore/remove-output.value-from-tests

**Comment by Owlfred** [comment]
End to end tests: :hourglass_flowing_sand: **STARTING**

File: substra/sdk/backends/local/compute/worker.py [hunk 8]
```diff
@@ -302,6 +302,10 @@ def schedule_task(self, task: models.Task):
                         datasample_input_refs.append(task_input)
                         datasamples.append(asset)
 
+                        # In hybrid mode, weuse DataSample assets with a path equals to `None` as it created as fake
```
Comment (SdgJlbl) [inline substra/sdk/backends/local/compute/worker.py:305]:
```suggestion
                        # In hybrid mode, we use DataSample assets with a path equals to `None` as it created as fake
```

File: CHANGELOG.md (hunk 2 above)
Comment (SdgJlbl) [reply CHANGELOG.md:12]: maybe just "added `inputAsset` blablaba" 
I didn't notice that it was part of the "added" section at the first glance, so I was a bit confused. 
Might be worth mentioning too that it's mirroring the backend schema 🤷

**Review by SdgJlbl — APPROVED** [review]
Thanks a lot

**Pull request merged**